core/modules/filter/filter.pages.inc

```diff
@@ -491,28 +491,169 @@ function filter_format_editor_dialog_save($form, &$form_state) {
+  // Will add a default protocol if missing on an external url.
+  $url = _filter_create_valid_external_url($url);
+
+  if (_filter_url_is_external($url)) {
+    return $url;
+  }
+
+  // Urlencode an anchor as it's more likely to be typed in with unicode.
+  if (substr($url, 0, 1) == '#' && preg_match('/^(#)([\S]+)$/', $url, $matches)) {
+    return $matches['1'] . backdrop_encode_path($matches['2']);
+  }
+
+  // Break up the components so we can build up an aliased internal path.
+  $url_parts = backdrop_parse_url($url);
+
+  // Lookup valid internal paths if multilingual
+  if (module_exists('locale') && language_multilingual()) {
+    if ($language = _filter_lookup_locale_path($url)) {
+      $url_parts['language'] = $language;
+    }
+  }
+
+  return url($url_parts['path'], $url_parts);
+}
+
+/**
+ * Forms a valid external URL if possible.
+ *
+ * Trims whitespace and automatically adds an http:// to addresses that
+ * appear to be external URLs but have no specified protocol.
+ *
+ * @param string $url
+ *   The URL entered by the user.
+ *
+ * @return string
+ *   The URL with the protocol prepended if needed.
+ *
+ * @see link_cleanup_url().
+ */
+function _filter_create_valid_external_url($url) {
+  $url = trim($url);
+
+  // Check if there is no protocol specified.
+  $protocol_match = preg_match('/^([a-z0-9][a-z0-9\.\-_]*:\/\/)/i', $url);
+  if (empty($protocol_match)) {
+    // But should there be? Add an automatic http:// if it starts with a domain name.
+    $domain_match = preg_match('/^(([a-z0-9]([a-z0-9\-_]*\.){1,63})([a-z]{2,63}))/i', $url);
+    if (!empty($domain_match)) {
+      $url = "http://" . $url;
+    }
+  }
+
+  return $url;
+}
+
+/**
+ * Checks if a URL is external to Backdrop.
+ *
+ * Checks for a protocol as a sign that the URL is external. We want to include
+ * all external URLs even if they have dangerous protocols.
+ *
+ * @param string $url
+ *
+ * @return bool
+ *   TRUE if URL is external.
+ *
+ * @see url_is_external()
+ */
+function _filter_url_is_external($url) {
+  $protocol_match = preg_match('/^([a-z0-9][a-z0-9\.\-_]*:\/\/)/i', $url);
+  if (!empty($protocol_match)) {
+    return TRUE;
+  }
+  return FALSE;
+}
+
+/**
+ * Given a URL, find an alias and matching lanaguage.
+ *
+ * @param $url
+ *   The path to investigate for corresponding aliases.
+ *
+ * @return object|FALSE
+ *   The langcode.
+ */
+function _filter_lookup_locale_path($url) {
+  $languages = language_list(TRUE);
+    foreach ($languages as $language) {
+      $alias = backdrop_lookup_path('alias', $url, $language->langcode);
+      if ($alias) {
+        return $language;
+      }
+    }
+  return NULL;
+}
+
+/**
+ * Check for dangerous protocols.
+ *
+ * If it has a protocol that doesn't match any of the allowed protocols then
+ * flag it as dangerous.
+ *
+ * @param string $url
+ *
+ * @return bool
+ *   True if any dangerous protocols present.
+ */
+function _filter_check_dangerous_protocols($url) {
```

Comment:
I wonder, why function backdrop_strip_dangerous_protocols() wasn't sufficient to accomplish that task.

Function check_url() uses that, for example.

Reply:
That does work to strip protocols but I was going by your tests which was checking the validity of dangerous protocols so I needed to *not* strip the protocol but just give the user a warning.

Reply:
One option would be to run the url through backdrop_strip_dangerous_protocols() and compare the outcome with the original url. Then there'd be no need for *yet* another regex pattern...

Reply:
I've done this.